mm2src/coins/utxo/rpc_clients.rs

```diff
@@ -1678,6 +1707,9 @@ impl ElectrumClientImpl {
 
     /// Get available protocol versions.
     pub fn protocol_version(&self) -> &OrdRange<f32> { &self.protocol_version }
+
+    /// Get block headers storage.
+    pub fn block_headers_storage(&self) -> &Option<BlockHeaderStorage> { &self.block_headers_storage }
```

Comment:
It seems worth making `BlockHeaderStorage` non-optional. I guess we would like to enable SPV proof validation for as many coins as possible, so initializing storage by default makes sense. It should also ease the handling in various places, avoiding `match` and `if let Some(_) = ...`.

Reply:
There are multiple implications to this.

First is storage requirements for mobile and web DEXes, although bitcoin block headers consumes space in the range of 50MB, that's not the case for other UTXO coins that have a lot less blocktimes, to solve this we can just keep the last few validated headers and remove the rest since we wouldn't need them to validate swaps or channels transactions after some time is passed.

The second problem is that all block headers needs to be downloaded before validating a transaction the first time the app is used, as a solution we can provide a validated block header information that is recent in the coins config that we can start the download from as we discussed before, but we would need to update this in the coins config periodically. Also, when validating a transaction we have to wait for block headers sync first, it would be ideal to sync block headers when enabling a coin but this can be done only in the coins that use task manager in enabling them.

In conclusion, making `BlockHeaderStorage` non-optional might require a lot of code changes and still requires more thinking about it's implications, so it might be best to move this to the next sprint and not make it part of this PR since SPV validation is not finished yet as I need to use the difficulty calculations in validating headers also.

Reply:
> Also, when validating a transaction we have to wait for block headers sync first, it would be ideal to sync block headers when enabling a coin but this can be done only in the coins that use task manager in enabling them.

If SPV proof validation is disabled, we can still initialize the storage without starting sync and using the storage space. It might be confusing, though, but will still allow avoiding `if let Some(storage)` in many places - it is just an overhead because storage *must* be initialized if SPV proof validation is turned on. There should be a way to handle it without an `Option`. But yeah, you can do it on the next sprint 🙂 

> coins config that we can start the download from as we discussed before, but we would need to update this in the coins config periodically.

I might be wrong, but I think I didn't mention that we need to download block headers partially. I meant to have a specific block difficulty as anchor for PoW validation, but still have all headers in DB.

Reply:
> But yeah, you can do it on the next sprint 🙂

Added this to the issue's checklist https://github.com/KomodoPlatform/atomicDEX-API/issues/1045#issue-971642359